packages/server/graphql/mutations/loginWithGoogle.ts

```diff
@@ -0,0 +1,123 @@
+      type: new GraphQLNonNull(GraphQLID),
+      description: 'The code provided from the OAuth2 flow'
+    },
+    segmentId: {
+      type: GraphQLID,
+      description: 'optional segment id created before they were a user'
+    },
+    invitationToken: {
+      type: GraphQLID,
+      description: 'if present, the user is also joining a team'
+    }
+  },
+  resolve: rateLimit({perMinute: 50, perHour: 500})(
+    async (_source, {code, invitationToken, segmentId}, context: GQLContext) => {
+      const r = await getRethink()
+
+      // VALIDATION
+      const manager = await GoogleServerManager.init(code)
+      const {id} = manager
+      if (!id) {
+        return standardError(new Error('Invalid login code'))
+      }
+      const {email, picture, name, email_verified, sub} = id
+
+      const existingUser = await r
+        .table<User>('User')
+        .getAll(email, {index: 'email'})
+        .nth(0)
+        .default(null)
+        .run()
+
+      if (existingUser) {
+        const {id: viewerId, identities, rol} = existingUser
+        let googleIdentity = identities.find(
+          (identity) => identity.type === AuthIdentityTypeEnum.GOOGLE
+        ) as AuthIdentityGoogle
+        if (!googleIdentity) {
+          const [bestIdentity] = identities
+          if (!bestIdentity) {
+            return standardError(new Error('No identity found! Please contact support'))
+          }
+          const {type, isEmailVerified} = bestIdentity
+          // the existing account could belong to a squatter. If it's theirs, they need to verify the email
+          // if it's not, they need to reset the password
+          if (!isEmailVerified) {
+            if (type === AuthIdentityTypeEnum.LOCAL) {
+              return standardError(new Error('Try logging in with email and password'))
+            }
+            throw new Error(`Unknown identity type: ${type}`)
+          }
+
+          // add a google identity to the user if we're sure the local isn't a squatter
+          googleIdentity = new AuthIdentityGoogle({
+            isEmailVerified: email_verified !== 'false',
+            id: sub
+          })
+          identities.push(googleIdentity) // mutative
+          await r
+            .table('User')
+            .get(viewerId)
+            .update({
+              identities
+            })
+            .run()
+        }
+        // log them in
+        sendSegmentIdentify(viewerId).catch()
+        // MUTATIVE
+        context.authToken = new AuthToken({sub: viewerId, rol, tms: existingUser.tms})
+        return {
+          userId: viewerId,
+          // create a brand new auth token using the tms in our DB, not auth0s
+          authToken: encodeAuthToken(context.authToken)
+        }
+      }
+
+      // it's a new user!
+      const nickname = name || email.substring(0, email.indexOf('@'))
+      const preferredName = nickname.length === 1 ? nickname.repeat(2) : nickname
+      const userId = `u_${shortid.generate()}`
```

Comment:
See other comment: `parabol|${shortid.generate()}`. Maybe make userId generation a small utility function for DRY?